src/quick-lint-js/fe/parse.cpp

```diff
@@ -734,6 +734,81 @@ void parser::consume_semicolon() {
   }
 }
 
+void parser::error_on_pointless_nullish_coalescing_operator(
+    expression::binary_operator* ast) {
+  
+  auto is_nullish_operator = [](string8_view s) -> bool {
+    return s == u8"??"_sv;
+  };
+
+  for (span_size i = 0; i < ast->child_count() - 1; i++) {
+    source_code_span op_span = ast->operator_spans_[i];
+    if (is_nullish_operator(op_span.string_view())) {
+        if (i >= 1){
+            // lhs is a multi-child expression
+            return;
+        } else {
+            this->check_lhs_for_null_potential(ast->child(i)->without_paren(), op_span);
+        }
+    }
+  }
+}
+
+void parser::check_lhs_for_null_potential(expression* lhs,
+                                           source_code_span op_span) {
+  
+  auto binary_operator_is_never_null = [](string8_view s) -> bool {
+    // these 4 binary operators can resolve to a null value
+    string8_view can_resolve_to_null[4] = { u8"&&"_sv, u8"??"_sv, u8","_sv, u8"||"_sv };
+    for(int i = 0; i < 4; i++){
+        if(can_resolve_to_null[i] == s){
+            return false;
+        }
+    }
+    return true;
+  };
+  
+  switch(lhs->kind()){
+    case expression_kind::literal:
+        if (lhs->span().string_view() == u8"null"_sv){
+            return;
+        }
+        if (lhs->span().string_view() == u8"undefined"_sv){
+            return;
+        }
+        this->diag_reporter_->report(diag_pointless_nullish_coalescing_operator{
+          .question_question = op_span});
+        return;
+    case expression_kind::rw_unary_suffix:
+        this->diag_reporter_->report(diag_pointless_nullish_coalescing_operator{
+          .question_question = op_span});
+        return;
+    case expression_kind::unary_operator:{
+        auto* maybe_void_lhs = static_cast<expression::unary_operator*>(lhs);
+        if (maybe_void_lhs->is_void_operator() == false){
+            this->diag_reporter_->report(diag_pointless_nullish_coalescing_operator{
+            .question_question = op_span});
+        }
+        return;
+        }
+    case expression_kind::_typeof:
+        this->diag_reporter_->report(diag_pointless_nullish_coalescing_operator{
+            .question_question = op_span});
+        return;
+    case expression_kind::binary_operator:{
+        auto* operator_lhs = static_cast<expression::binary_operator*>(lhs);
+        source_code_span binary_op_span = operator_lhs->operator_spans_[0];
+        if (binary_operator_is_never_null(binary_op_span.string_view())){
```

Comment:
I will try to look into this, but I'm not entirely sure how to differentiate the main operator of the expression and the sub-operators of the left and right operands. I checked expression.h and there doesn't seem to already be a class method to return this. Are there any existing helper functions that do this? Or do you think it would be a good solution to write a helper function that returns the lowest priority operator, and breaking ties based on which is the leftmost operator?

Reply:
> I'm not entirely sure how to differentiate the main operator of the expression and the sub-operators of the left and right operands.

Yeah. quick-lint-js hasn't needed to deal with precedence much before, so the support for it is weak.

> Are there any existing helper functions that do this?

Not right now. I do have an idea for how it'd look though; see #557.

> So i'm thinking that it might just be enough to return false if 1 of the operators are one of these 4 && ?? , ||

Sounds good!

If you can't find a reliable method (without writing lots of code), then I say don't worry about it for now and just ignore binary operators to the left of `??`.